core/node/base_token_fetcher/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+use std::{cmp::min, sync::Arc};
+
+use anyhow::Context;
+use async_trait::async_trait;
+use hex::ToHex;
+use metrics::atomics::AtomicU64;
+use tokio::sync::Mutex;
+use zksync_config::configs::BaseTokenFetcherConfig;
+
+/// Trait used to query the stack's native token conversion rate. Used to properly
+/// determine gas prices, as they partially depend on L1 gas prices, denominated in `eth`.
+#[async_trait]
+pub trait ConversionRateFetcher: 'static + std::fmt::Debug + Send + Sync {
+    fn conversion_rate(&self) -> anyhow::Result<u64>;
```

Comment:
IIRC last time we've discussed it we came to a conclusion that `u64` probably won't work well here.
Is it going to be tackled in some other PR?

Reply:
The work for that was started on [this PR](https://github.com/matter-labs/zksync-era/pull/1324) but then got frozen after some discussions.

Reply:
@popzxc @jrchatruc we'll 
1. release this with the current limitations but should be very clear about them in docs and comments
2. go with `U256` as a follow-up

For now, I'll take care of docs and @jrchatruc please add comment/s where you see fit.

Reply:
I think `u64/U256` issue was about the type used for gas prices (and potential of overflow).
Here, however, we're discussing the fact that we cannot adequately express token prices (e.g. if the token price is $2000, and ETH price is $3000, we cannot express it using `u64`. Similarly, we cannot _precisely_ express token costing $0.7, creating some rounding error, and I don't know if it's OK).

Reply:
You're right, I got things mixed up, but we did make a PR which ultimately got merged into the `U256` one addressing the fractionary conversion rate issue ([here's a comment about it](https://github.com/matter-labs/zksync-era/pull/1324#issuecomment-2000339914) on the `U256` PR). I think in all the different PRs it got lost

Reply:
@popzxc Good point apologies for the confusion. I see what you're saying given this is used simply as `self.bound_gas_price(calculated_price) * conversion_rate`, and the linked PR also has nothing to do with this comment.

Reply:
@jrchatruc So just to be clear, it will be changed to correctly represent a ratio in this PR, right?

Reply:
We could definitely do that; for that we would need to merge [this PR](https://github.com/lambdaclass/zksync-era/pull/219) into this branch first before merging this one into `main`.

Reply:
@jrchatruc looking at the PR title alone (the diff is broken there so I cannot see exact changes), it seems to be off-topic. This PR seems to be again about porting gas prices from `u64` to `U256` which AFAIK we decided not to do for now, since it has some issues. What we need to do is to change the data type returned by `conversion_rate` from `u64` to, for example, [Ratio](https://docs.rs/num/latest/num/rational/struct.Ratio.html) in order to precisely represent relations between ETH and native token prices.

Reply:
+1 to @popzxc

Reply:
We have creates a [PR](https://github.com/lambdaclass/zksync-era/pull/220) that moves the data type returned by `conversion_rate` to `BigDecimal`, we think it's the best approach, yet we also experimented a bit with `Ratio` and got it to work too.